python/src/uagents/agent.py

```diff
@@ -880,6 +882,46 @@ async def _process_message_queue(self):
             )
             if handler is None:
                 if not is_user_address(sender):
+                    for protocol in context.protocols.values():
+                        if hasattr(protocol, "rules") and protocol.is_included(
+                            schema_digest
+                        ):
+                            state = protocol.get_current_state(session)
+                            context.logger.debug(
+                                "current state: "
+                                f"{(protocol.models[state].__name__ if state else 'n/a')}"
+                            )
+                            is_valid = protocol.is_valid_message(session, schema_digest)
+                            context.logger.debug(f"message allowed: {is_valid}")
+
+                            if not is_valid:
+                                context.session = None
+                                await _send_error_message(
+                                    context,
+                                    sender,
+                                    ErrorMessage(
+                                        error="Unexpected message in dialogue"
+                                    ),
+                                )
+                                continue
+
+                            if protocol.is_starter(schema_digest):
+                                self._ctx.logger.debug("dialogue started")
+                            elif protocol.is_ender(schema_digest):
+                                self._ctx.logger.debug(
+                                    "dialogue ended, cleaning up session"
+                                )
+                            else:
+                                self._ctx.logger.debug("dialogue picked up")
+
+                            context.dialogue = protocol.get_session(session)
+                            protocol.add_message(
+                                session_id=session,
+                                message_type=self._models[schema_digest].__name__,
+                                sender=sender,
+                                receiver=self.address,
+                                content=message,
+                            )
```

Comment:
Feels like this block should go in a dedicated function somewhere, maybe in `Dialogue`?

Reply:
Good point, will be transferred to `Dialogue`.

Reply:
After a bit of consideration I think it would be good to keep this code in `agent.py` because about half of the lines is just debugging messages. The rest accesses the context or agent and moving it to the `Dialogue` class would add unnecessary complexity.
We can remove the debug messages later or add additional verbosity throughout the framework but in principle I agree that it otherwise creates a logging imbalance.